mono/mini/method-to-ir.c

```diff
@@ -6985,53 +6985,35 @@ is_jit_optimizer_disabled (MonoMethod *m)
 }
 
 static gboolean
-is_supported_tail_call (MonoCompile *cfg, MonoMethod *method, MonoMethod *cmethod, MonoMethodSignature *fsig, int call_opcode)
+is_supported_tail_call (MonoCompile *cfg, MonoMethod *method, MonoMethod *cmethod, MonoMethodSignature *fsig, int call_opcode, gboolean virtual_, MonoInst *vtable_arg)
 {
-	gboolean supported_tail_call;
 	int i;
 
-	supported_tail_call = mono_arch_tail_call_supported (cfg, mono_method_signature (method), mono_method_signature (cmethod));
+	g_assertf (call_opcode == CEE_CALL || call_opcode == CEE_CALLVIRT || call_opcode == CEE_CALLI, "%s (%d)", mono_opcode_name (call_opcode), (int)call_opcode);
+
+	if (       (fsig->hasthis && cmethod->klass->valuetype)  // This might point to the current method's stack. Emit range check?
+		|| (cmethod->flags & METHOD_ATTRIBUTE_PINVOKE_IMPL)
```

Comment:
This is a much harder to read than the previous code, and it's shuffling around the code for no benefits.